presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -2069,15 +2094,19 @@ else if (partitionUpdate.getUpdateMode() == NEW || partitionUpdate.getUpdateMode
                         partitionUpdate.getStatistics(),
                         columnTypes,
                         getColumnStatistics(partitionComputedStatistics, partition.getValues()));
-                metastore.addPartition(
-                        session,
-                        handle.getSchemaName(),
-                        handle.getTableName(),
-                        table.getStorage().getLocation(),
-                        false,
-                        partition,
-                        partitionUpdate.getWritePath(),
-                        partitionStatistics);
+
+                // New partition or overwriting existing partition by staging and moving the new partition
+                if (!existingPartition || (existingPartition && handle.getLocationHandle().getWriteMode() != DIRECT_TO_TARGET_EXISTING_DIRECTORY)) {
```

Comment:
This looks like a bug? You mean
```java
if (!existingPartition || partitionUpdate.getUpdateMode() == OVERWRITE)
```
?

Reply:
If partitionUpdate.getUpdateMode() != OVERWRITE, then we will throw error on line 2089.  If we just check for (partitionUpdate.getUpdateMode() == OVERWRITE) here, even when writeMode is DIRECT_TO_TARGET_EXISTING_DIRECTORY, we will add partition, which is wrong. Since for DIRECT_TO_TARGET_EXISTING_DIRECTORY, the partition already exists and we are removing old files. I simplified the condition to if (!existingPartition || handle.getLocationHandle().getWriteMode() != DIRECT_TO_TARGET_EXISTING_DIRECTORY)

Reply:
If that is the case, we can still simlify the condition to be
```java
if (!existingPartition || handle.getLocationHandle().getWriteMode() != DIRECT_TO_TARGET_EXISTING_DIRECTORY)
```
?